### ISSUE TITLE
What is categorical loss?

### ISSUE THREAD

**Issue by zhihanyang2022** [jariasf/GMVAE#1]
In `/pytorch/model/GMVAE.py` line 77, I see that the categorical loss, as defined in the file, is the negative entropy minus some constant. I find this confusing for two reasons:

- This loss term is not mentioned in the original paper of the Gumbel-Softmax distribution (https://arxiv.org/pdf/1611.01144.pdf).

- The constant term at the end of this loss can be omitted, right? It is a constant and it does not depend on network parameters. What is the motivation for this term?

I hope some comments can be added to this line of code. I can also help, too!

**Comment by jariasf** [comment]
Sorry if the name of the loss was misleading. I called it categorical loss because the terms involved in this loss where based on the categorical latent variable that represents the clusters.

- _This loss term is not mentioned in the original paper of the Gumbel-Softmax distribution (https://arxiv.org/pdf/1611.01144.pdf)._

Yes, it is not mentioned in the original paper. You can find the loss in the categorical VAE that the authors implemented as an application of the Gumbel-Softmax distribution: https://blog.evjang.com/2016/11/tutorial-categorical-variational.html.

- _The constant term at the end of this loss can be omitted, right? It is a constant and it does not depend on network parameters. What is the motivation for this term?_

Yes, in terms of optimization it can be omitted because it is just a constant. This constant appears due to the ELBO derived from the probabilistic model. In the following picture you can see the inference and generative models used in this implementation. It is a modification of the M2 model proposed by Kingma et al. (https://arxiv.org/pdf/1406.5298):

![image](https://drive.google.com/uc?id=1uyPOldn8lekiiie5cy8yQRMNRI2DRs8D)

If you derive the ELBO of this model you will get something like this:

![image](https://user-images.githubusercontent.com/4966338/92984879-4a1c3380-f473-11ea-9139-5e25ba6a8bfa.png)

Here you can see the different distributions you have to implement:

- p(x|z) is a bernoulli or gaussian distribution depending on the type of data you are working with.
- p(z|y) is a gaussian distribution that depends on the categorical variable y. N(z|mean(y), var(y))
- p(y) is a uniform prior. In this implementation we are using MNIST (10 categories), that is why, the log of this prior is equal to log(1/10).
- q(z|x,y) is a gaussian distribution that depends on the input data x and categorical variable y. N(z|mean(x,y), var(x,y)).
- q(y|x) is a categorical distribution used in the entropy due to the expectation (E[log q(y|x)]).

You can also group some term and derive an ELBO with KL-divergences for both the gaussian and categorical distributions.

**Comment by zhihanyang2022** [comment]
Thanks a lot for the detailed response! I'm still in the process of understanding the derivation of the ELBO. Are there any helpful resources that I should consult?

**Comment by ZhangZhenmiao** [comment]
Thanks a lot! After reading the response, I am still confused about the categorical loss. In /pytorch/model/GMVAE.py line 68, the prob is the softmax of logits, therefore it seems the categorical loss is derived from logits and its softmax. So does that mean the softmax represents the prior prob_cat?

**Comment by nikhilrayaprolu** [comment]
@jariasf Can you point where the categorical loss is mentioned in the blog? I could find any extra terms other than the ELBO in the implementation.
https://blog.evjang.com/2016/11/tutorial-categorical-variational.html

**Comment by jariasf** [comment]
@nikhilrayaprolu On the blog the categorical loss is given in the following line of code:

```
# loss and train ops
kl_tmp = tf.reshape(q_y*(log_q_y-tf.log(1.0/K)),[-1,N,K])
KL = tf.reduce_sum(kl_tmp,[1,2])
...
```
On the blog, as well as in the paper, this term is not called categorical loss. I called it categorical loss because the terms involved in this loss were based on the categorical latent variable that represents the clusters.

**Comment by jariasf** [comment]
I'm closing this issue due to inactivity
